homeassistant/components/remote/__init__.py

```diff
@@ -0,0 +1,151 @@
+
+For more details about this component, please refer to the documentation
+at https://home-assistant.io/components/remote/
+"""
+from datetime import timedelta
+import logging
+import os
+
+import voluptuous as vol
+from homeassistant.config import load_yaml_config_file
+from homeassistant.helpers.entity_component import EntityComponent
+from homeassistant.helpers.entity import ToggleEntity
+import homeassistant.helpers.config_validation as cv
+from homeassistant.const import (
+    STATE_ON, SERVICE_TURN_ON, SERVICE_TURN_OFF, ATTR_ENTITY_ID)
+from homeassistant.components import group
+from homeassistant.helpers.config_validation import PLATFORM_SCHEMA  # noqa
+
+ATTR_DEVICE = 'device'
+ATTR_COMMAND = 'command'
+ATTR_ACTIVITY = 'activity'
+SERVICE_SEND_COMMAND = 'send_command'
+SERVICE_SYNC = 'sync'
+ATTR_DEFAULT = ''
+
+DOMAIN = 'remote'
+SCAN_INTERVAL = 30
+
+GROUP_NAME_ALL_REMOTES = 'all remotes'
+ENTITY_ID_ALL_REMOTES = group.ENTITY_ID_FORMAT.format('all_remotes')
+
+ENTITY_ID_FORMAT = DOMAIN + '.{}'
+
+MIN_TIME_BETWEEN_SCANS = timedelta(seconds=10)
+
+REMOTE_SERVICE_SCHEMA = vol.Schema({
+    vol.Optional(ATTR_ENTITY_ID): cv.entity_ids,
+    vol.Optional(ATTR_DEVICE): cv.string,
+    vol.Optional(ATTR_COMMAND): cv.string,
+    vol.Optional(ATTR_ACTIVITY): cv.string
+})
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def is_on(hass, entity_id=None):
+    """Return if the remote is on based on the statemachine."""
+    entity_id = entity_id or ENTITY_ID_ALL_REMOTES
+    return hass.states.is_state(entity_id, STATE_ON)
+
+
+def turn_on(hass, activity=None, entity_id=None):
+    """Turn all or specified remote on."""
+    data = {ATTR_ACTIVITY: activity}
+    if entity_id:
+        data[ATTR_ENTITY_ID] = entity_id
+    hass.services.call(DOMAIN, SERVICE_TURN_ON, data)
+
+
+def turn_off(hass, entity_id=None):
+    """Turn all or specified remote off."""
+    data = {ATTR_ENTITY_ID: entity_id} if entity_id else None
+    hass.services.call(DOMAIN, SERVICE_TURN_OFF, data)
+
+
+def send_command(hass, device=None, command=None, entity_id=None):
+    """Send a command to a device."""
+    data = {ATTR_DEVICE: str(device), ATTR_COMMAND: command}
+    if entity_id:
+        data[ATTR_ENTITY_ID] = entity_id
+    hass.services.call(DOMAIN, SERVICE_SEND_COMMAND, data)
+
+
+def sync(hass, entity_id=None):
+    """Sync remote device."""
+    data = {ATTR_ENTITY_ID: entity_id} if entity_id else None
+    hass.services.call(DOMAIN, SERVICE_SYNC, data)
+
+
+def setup(hass, config):
+    """Track states and offer events for remotes."""
+    component = EntityComponent(
+        _LOGGER, DOMAIN, hass, SCAN_INTERVAL, GROUP_NAME_ALL_REMOTES)
+    component.setup(config)
+
+    def handle_remote_service(service):
+        """Handle calls to the remote services."""
+        target_remotes = component.extract_from_service(service)
+
+        activity_id = service.data.get(ATTR_ACTIVITY, ATTR_DEFAULT)
+        device = str(service.data.get(ATTR_DEVICE, ATTR_DEFAULT))
```

Comment:
Defaults should be handled in the schema. Conversion is already handled in the schema. 

Also defaults shouldn't be an empty string. It's up to the platform to decide how to deal with the absence of a parameter.